### PR TITLE
Create2 Opcode

### PR DESCRIPTION
**Pull request by smcsicardi**
PR to implement the CREATE2 opcode

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 30 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 14 minor
* ![INFO][INFO] 7 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [OpCodes.java#L621](https://github.com/rsksmart/rskj/blob/ab5344a13f2ed20584363260cb4c3884bcc7f36c/rskj-core/src/main/java/org/ethereum/vm/OpCodes.java#L621): Explicitly declare the visibility for "OP_CREATE2". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2039)
1. ![CRITICAL][CRITICAL] [VM.java#L1381](https://github.com/rsksmart/rskj/blob/ab5344a13f2ed20584363260cb4c3884bcc7f36c/rskj-core/src/main/java/org/ethereum/vm/VM.java#L1381): Missing curly brace. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)
1. ![CRITICAL][CRITICAL] [VM.java#L1396](https://github.com/rsksmart/rskj/blob/ab5344a13f2ed20584363260cb4c3884bcc7f36c/rskj-core/src/main/java/org/ethereum/vm/VM.java#L1396): Missing curly brace. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)
1. ![MAJOR][MAJOR] [VM.java#L539](https://github.com/rsksmart/rskj/blob/ab5344a13f2ed20584363260cb4c3884bcc7f36c/rskj-core/src/main/java/org/ethereum/vm/VM.java#L539): Method org.ethereum.vm.VM.doSHL() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L554](https://github.com/rsksmart/rskj/blob/ab5344a13f2ed20584363260cb4c3884bcc7f36c/rskj-core/src/main/java/org/ethereum/vm/VM.java#L554): Method org.ethereum.vm.VM.doSHR() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L569](https://github.com/rsksmart/rskj/blob/ab5344a13f2ed20584363260cb4c3884bcc7f36c/rskj-core/src/main/java/org/ethereum/vm/VM.java#L569): Method org.ethereum.vm.VM.doSAR() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L1366](https://github.com/rsksmart/rskj/blob/ab5344a13f2ed20584363260cb4c3884bcc7f36c/rskj-core/src/main/java/org/ethereum/vm/VM.java#L1366): Method org.ethereum.vm.VM.doCREATE() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L1394](https://github.com/rsksmart/rskj/blob/ab5344a13f2ed20584363260cb4c3884bcc7f36c/rskj-core/src/main/java/org/ethereum/vm/VM.java#L1394): Method org.ethereum.vm.VM.doCREATE2() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [Program.java#L308](https://github.com/rsksmart/rskj/blob/ab5344a13f2ed20584363260cb4c3884bcc7f36c/rskj-core/src/main/java/org/ethereum/vm/program/Program.java#L308): Method org.ethereum.vm.program.Program.memorySave(DataWord, DataWord) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MINOR][MINOR] [OpCodes.java#L136](https://github.com/rsksmart/rskj/blob/ab5344a13f2ed20584363260cb4c3884bcc7f36c/rskj-core/src/main/java/org/ethereum/vm/OpCodes.java#L136): Move this variable to comply with Java Code Conventions. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1213)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 27 issues
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 14 minor
* ![INFO][INFO] 7 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [VM.java#L539](https://github.com/rsksmart/rskj/blob/a697ed158fb1ebe9c58afc04a11ed1935f16a220/rskj-core/src/main/java/org/ethereum/vm/VM.java#L539): Method org.ethereum.vm.VM.doSHL() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L554](https://github.com/rsksmart/rskj/blob/a697ed158fb1ebe9c58afc04a11ed1935f16a220/rskj-core/src/main/java/org/ethereum/vm/VM.java#L554): Method org.ethereum.vm.VM.doSHR() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L569](https://github.com/rsksmart/rskj/blob/a697ed158fb1ebe9c58afc04a11ed1935f16a220/rskj-core/src/main/java/org/ethereum/vm/VM.java#L569): Method org.ethereum.vm.VM.doSAR() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L1366](https://github.com/rsksmart/rskj/blob/a697ed158fb1ebe9c58afc04a11ed1935f16a220/rskj-core/src/main/java/org/ethereum/vm/VM.java#L1366): Method org.ethereum.vm.VM.doCREATE() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L1394](https://github.com/rsksmart/rskj/blob/a697ed158fb1ebe9c58afc04a11ed1935f16a220/rskj-core/src/main/java/org/ethereum/vm/VM.java#L1394): Method org.ethereum.vm.VM.doCREATE2() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [Program.java#L308](https://github.com/rsksmart/rskj/blob/a697ed158fb1ebe9c58afc04a11ed1935f16a220/rskj-core/src/main/java/org/ethereum/vm/program/Program.java#L308): Method org.ethereum.vm.program.Program.memorySave(DataWord, DataWord) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MINOR][MINOR] [OpCodes.java#L136](https://github.com/rsksmart/rskj/blob/a697ed158fb1ebe9c58afc04a11ed1935f16a220/rskj-core/src/main/java/org/ethereum/vm/OpCodes.java#L136): Move this variable to comply with Java Code Conventions. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1213)
1. ![MINOR][MINOR] [OpCodes.java#L140](https://github.com/rsksmart/rskj/blob/a697ed158fb1ebe9c58afc04a11ed1935f16a220/rskj-core/src/main/java/org/ethereum/vm/OpCodes.java#L140): Move this variable to comply with Java Code Conventions. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1213)
1. ![MINOR][MINOR] [OpCodes.java#L144](https://github.com/rsksmart/rskj/blob/a697ed158fb1ebe9c58afc04a11ed1935f16a220/rskj-core/src/main/java/org/ethereum/vm/OpCodes.java#L144): Move this variable to comply with Java Code Conventions. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1213)
1. ![MINOR][MINOR] [OpCodes.java#L621](https://github.com/rsksmart/rskj/blob/a697ed158fb1ebe9c58afc04a11ed1935f16a220/rskj-core/src/main/java/org/ethereum/vm/OpCodes.java#L621): Move this variable to comply with Java Code Conventions. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1213)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rskj-core/src/main/java/co/rsk/core/RskAddress.java [hunk 1]
```diff
@@ -116,4 +116,8 @@ public int hashCode() {
     public String toString() {
         return Hex.toHexString(bytes);
     }
+
+    public int length() {
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/core/RskAddress.java]: I think it's better to not expose this method and use the length of the `co.rsk.core.RskAddress#getBytes` array. This is because there is an special address in our domain, which is the Remasc sender, who has length equals to 1. It shouldn't be the case for this new opcode to be used by that address, but for keeping consistency along all possible RskAddress in our domain it's better to get the address length in the way I mentioned before.

**Comment by smcsicardi** [comment]
> It looks really great, I would take a look a the minor things I marked here an there. Thank you for keeping your commits so clean

Awesome, I'm fixing those things right now and will push them when it's ready

File: rskj-core/src/main/java/org/ethereum/crypto/HashUtil.java [hunk 2]
```diff
@@ -116,6 +117,30 @@
...
+     * sha3(0xff ++ msg.sender ++ salt ++ sha3(init_code)))[12:]
+     *
+     * @param senderAddr - creating address
+     * @param initCode - contract init code
+     * @param salt - salt to make different result addresses
+     * @return new address
+     */
+    public static byte[] calcSaltAddr(RskAddress senderAddr, byte[] initCode, byte[] salt) {
+        // 1 - 0xff length, 32 bytes - keccak-256
+        byte[] data = new byte[1 + senderAddr.getBytes().length + salt.length + 32];
```
[the remark this answers is not in the document]
Comment (smcsicardi) [reply rskj-core/src/main/java/org/ethereum/crypto/HashUtil.java]: This is changed, I left a comment explaining better each number:
We have
- 1 byte for the 0xff
- 32 bytes of the sha3(init_code)
- The sender address length
- The salt lenght

File: rskj-core/src/main/java/org/ethereum/vm/program/Program.java [hunk 3]
```diff
@@ -454,26 +470,46 @@ public void createContract(DataWord value, DataWord memStart, DataWord memSize)
...
         }
 
 
         // [5] COOK THE INVOKE AND EXECUTE
+        ProgramResult programResult = getProgramResult(senderAddress, nonce, value, contractAddress, endowment, programCode, gasLimit, track, newBalance);
+        if (programResult == null) {
+            return;
+        }
+
+        // 5. REFUND THE REMAIN GAS
```
[the remark this answers is not in the document]
Comment (smcsicardi) [reply rskj-core/src/main/java/org/ethereum/vm/program/Program.java]: These numbers were like this since the master branch, I removed the numbers that didnt made sense

File: rskj-core/src/test/java/co/rsk/vm/VMExecutionTest.java [hunk 4]
```diff
@@ -656,6 +665,155 @@ public void returnDataCopyBasicGasCost() {
...
+    }
+
+    private void callCreate2(String address, String salt, String pushInitCode, int size, int intOffset, String expected, long gasExpected) {
+        ActivationConfig.ForBlock activationConfig = mock(ActivationConfig.ForBlock.class);
+        when(activationConfig.isActive(RSKIP125)).thenReturn(true);
+        callCreate2WithActivationConfig(address, salt, pushInitCode, size, intOffset, expected, gasExpected, activationConfig);
+    }
+
+    @Test
+    public void testCREATE2_0() {
```
[the remark this answers is not in the document]
Comment (smcsicardi) [reply rskj-core/src/test/java/co/rsk/vm/VMExecutionTest.java]: Tests are now much better explained, each with a specific purpose

File: rskj-core/src/test/java/org/ethereum/util/HashUtilTest.java [hunk 5]
```diff
@@ -103,4 +104,69 @@ public void testRIPEMD160_Multiple() {
         String result2 = Hex.toHexString(HashUtil.ripemd160("test2".getBytes()));
         assertEquals(expected2, result2);
     }
+
+    @Test
+    public void testCalcSaltAddr0() {
```
[the remark this answers is not in the document]
Comment (smcsicardi) [reply rskj-core/src/test/java/org/ethereum/util/HashUtilTest.java]: Same as before, tests are now better explained

File: rskj-core/src/main/java/org/ethereum/vm/program/Program.java [hunk 6]
```diff
@@ -454,26 +470,46 @@ public void createContract(DataWord value, DataWord memStart, DataWord memSize)
         Repository track = getStorage().startTracking();
 
         //In case of hashing collisions, check for any balance before createAccount()
-        if (track.isExist(newAddress)) {
-            Coin oldBalance = track.getBalance(newAddress);
-            track.createAccount(newAddress);
-            track.addBalance(newAddress, oldBalance);
+        if (track.isExist(contractAddress)) {
```
Comment (juli) [inline rskj-core/src/main/java/org/ethereum/vm/program/Program.java]: for CREATE2 collisions are possible see EIP-104
 * the code and nonce of the existing address must be checked and fail if non zero
 * the nonce of the `newAddress` must be incremented

Comment (smcsicardi) [reply rskj-core/src/main/java/org/ethereum/vm/program/Program.java]: Ok, so the nonce increase of the `newAdress` would mean changing behavior of the `CREATE` opcode, because each new contract created would start with a non-zero nonce, we talked with @SergioDemianLerner and he decided it was better to keep compatibility with Ethereum so we are going to add the nonce increase, but only if the RSKIP is activated

**Comment by smcsicardi** [comment]
To sum up the new changes:

- The increaseNonce was added for the `newAddress` contract, this is for both `CREATE` and `CREATE2` opcodes, so with this RSKIP the `CREATE` opcode behaves differently than before.

- Because contracts created with this opcodes now start with nonce 1, each new address creates is different than before, so it should be noted that old .json tests had to be modified because of this change, which Ethereum also changed some time ago, not only changing the address but also checking that the old address doesn't exist (they have a `"shouldnotexist"` field on their tests Fillers to check this).
 
- Address Collision: Now there is a check that doesn't allow this to happen if the RSKIP is active, we set an exception to the `programResult` and set the programCode to empty so it doesn't executes and writes over the exception. 

TL,DR: This RSKIP changes CREATE behavior to improve compatibility with Ethereum which made the change some time ago in [EIP-161 (section A.)](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-161.md)

File: rskj-core/src/main/java/org/ethereum/vm/program/Program.java [hunk 7]
```diff
@@ -461,32 +478,74 @@ public void createContract(DataWord value, DataWord memStart, DataWord memSize)
...
+            // Hashing collisions in CREATE are rare, but in CREATE2 are possible
+            // We check for the nonce to be zero and that the code to be empty
+            // if any of this conditions fails, the contract creation fails:
+            // We set an exception as the program result and set the program
+            // to empty so it does not executes
+
+
+            byte[] code = track.getCode(contractAddress);
+
+            if (getActivations().isActive(ConsensusRule.RSKIP125) && ((code == null || code.length == 0) ||
```
Comment (juli) [inline rskj-core/src/main/java/org/ethereum/vm/program/Program.java]: " and the destination address already has either nonzero nonce, or *nonempty code*, then the creation throws *immediately*"

File: rskj-core/src/main/java/org/ethereum/vm/program/Program.java [hunk 8]
```diff
@@ -461,32 +478,74 @@ public void createContract(DataWord value, DataWord memStart, DataWord memSize)
...
+                track.getNonce(contractAddress).compareTo(BigInteger.ZERO) != 0)) {
+                programResult.setException(ExceptionHelper.addressCollisionException(contractAddress));
+                programCode = new byte[]{};
+            }
```
Comment (juli) [inline rskj-core/src/main/java/org/ethereum/vm/program/Program.java]: execution continues here why not aborting immediately?

Comment (smcsicardi) [reply rskj-core/src/main/java/org/ethereum/vm/program/Program.java]: Good question, the EIP contradicts itself a little in that part. On one hand it says it should throw immediately, on the other that it should fail...
> with exactly the same behavior as would arise if the first byte in the init code were an invalid opcode 

So what I did was force the same path that it would follow for this error.

Looking in different Ethereum implementations I saw that they approach this in different ways, ethereumJ does what we do (force the same error and continuing the execution), geth fails immediately and so does aleth. 

I think you are right and failing immediately is the right thing to do, will do some tests and upload the fixed implementation.

**Comment by smcsicardi** [comment]
Update: Discussing with @SergioDemianLerner he thought it was better to not consume all gas in the case of an address collision. The new code implements this change and its corresponding tests

**Comment by smcsicardi** [comment]
Update: A new RSKIP is being made for a specific corner case of the CREATE2, so to merge this we will wait for the new RSKIP to be implemented (WIP).

Basically it will prevent the creation of a contract in an address destroyed previously in a transaction to be included in the same block, so the check will have to be made in the cache of the block

File: rskj-core/src/main/java/co/rsk/core/TransactionExecutorFactory.java [hunk 9]
```diff
@@ -65,7 +69,7 @@ public TransactionExecutor newInstance(
             Repository track,
             Block block,
             long totalGasUsed,
-            boolean vmTrace) {
+            boolean vmTrace, Set<DataWord> deletedAccounts) {
```
Comment (lsebrie) [inline rskj-core/src/main/java/co/rsk/core/TransactionExecutorFactory.java]: Add extra line here please

File: rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java [hunk 10]
```diff
@@ -93,14 +92,15 @@
     private BigInteger mEndGas = BigInteger.ZERO;
     private long basicTxCost = 0;
     private List<LogInfo> logs = null;
+    private Set<DataWord> deletedAccounts;
```
Comment (lsebrie) [inline rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java]: Add a final here

File: rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java [hunk 11]
```diff
@@ -118,6 +118,7 @@ public TransactionExecutor(
         this.precompiledContracts = precompiledContracts;
         this.playVm = playVm;
         this.enableRemasc = remascEnabled;
+        this.deletedAccounts = new HashSet<>(deletedAccounts);
```
Comment (lsebrie) [inline rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java]: If you add final to the field you can remove hashset creation

File: rskj-core/src/test/java/co/rsk/test/World.java [hunk 12]
```diff
@@ -19,6 +19,7 @@
 package co.rsk.test;
 
 import co.rsk.config.TestSystemProperties;
+import co.rsk.core.RskAddress;
```
Comment (lsebrie) [inline rskj-core/src/test/java/co/rsk/test/World.java]: Remove this please

File: rskj-core/src/test/java/co/rsk/test/builders/TransactionBuilder.java [hunk 13]
```diff
@@ -55,6 +56,11 @@ public TransactionBuilder receiverAddress(byte[] receiverAddress) {
         return this;
     }
 
+    public TransactionBuilder receiverAddress(RskAddress receiverAddress) {
```
Comment (lsebrie) [inline rskj-core/src/test/java/co/rsk/test/builders/TransactionBuilder.java]: Remove this method it's not used

File: rskj-core/src/test/resources/dsl/createAfterSuicide/CreateCreator.sol [hunk 14]
```diff
@@ -0,0 +1,13 @@
+pragma solidity ^0.5.9;
+import "./Creator.sol";
+
+contract CreateCreator {
+    constructor() public{
+        bytes memory code = type(Creator).creationCode;
+        uint256 salt = 0;
+	    address addr;
```
Comment (lsebrie) [inline rskj-core/src/test/resources/dsl/createAfterSuicide/CreateCreator.sol]: Remove unnecesary spaces

File: rskj-core/src/test/resources/dsl/createAfterSuicide/Creator.sol [hunk 15]
```diff
@@ -0,0 +1,11 @@
+pragma solidity ^0.5.9;
+import "./TestChild.sol";
+
+contract Creator {
+    function run() public{
+        bytes memory code = type(TestChild).creationCode;
+        uint256 salt = 0;
+	    address addr;
```
Comment (lsebrie) [inline rskj-core/src/test/resources/dsl/createAfterSuicide/Creator.sol]: Remove unnecesary spaces
